src/TableOfContents/index.tsx

```diff
@@ -107,13 +109,26 @@ function TableOfContentsInner<T extends TableOfContentsItem = TableOfContentsIte
 
   // an array of functions. Invoking the N-th function toggles the expanded flag on the N-th content item
   const toggleExpandedFunctions = React.useMemo(() => {
-    return range(contents.length).map(i => () =>
-      setExpanded(current => ({
-        ...current,
-        [i]: !current[i],
-      })),
-    );
-  }, [contents.length]);
+    return range(contents.length).map(i => () => {
+      let childrenToCollapse = {};
+      if (expanded[i]) {
+        const item = contents[i];
+        const children = findDescendantIndices(item.depth ?? 0, i, contents.slice(i + 1));
+        childrenToCollapse = children.reduce((obj, index) => {
+          obj[index] = false;
+          return obj;
+        }, {});
+      }
+      setExpanded(current => {
+        return {
+          ...current,
+          [i]: !current[i],
+          ...childrenToCollapse,
+        };
+      });
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [contents, contents.length, expanded]);
```

Comment:
Oh this is one of the reasons why.
If you make the above change you'll only depend on `contents` and can get rid of the eslint inhibit.

Reply:
I think we still would need that `contents.length`, just `contents` would be enough to react on length change, right?